src/libDirectoryService/DSBlockMessage.h

```diff
@@ -0,0 +1,93 @@
+* and, to the extent permitted by law, all liability for your use of the code is disclaimed. 
+* Some programs in this code are governed by the GNU General Public License v3.0 (available at 
+* https://www.gnu.org/licenses/gpl-3.0.en.html) (‘GPLv3’). The programs that are governed by 
+* GPLv3.0 are those programs that are located in the folders src/depends and tests/depends 
+* and which include a reference to GPLv3 in their program files.
+**/
+
+#ifndef __DSBLOCKMESSAGE_H__
+#define __DSBLOCKMESSAGE_H__
+
+#include "libCrypto/Schnorr.h"
+#include "libNetwork/Peer.h"
+#include "libUtils/SanityChecks.h"
+#include <deque>
+#include <map>
+#include <utility>
+#include <vector>
+
+class ShardingStructure
+{
+    // Sharding structure message format:
+    // [4-byte num of committees]
+    // [4-byte committee size]
+    //   [33-byte public key] [16-byte ip] [4-byte port]
+    //   [33-byte public key] [16-byte ip] [4-byte port]
+    //   ...
+    // [4-byte committee size]
+    //   [33-byte public key] [16-byte ip] [4-byte port]
+    //   [33-byte public key] [16-byte ip] [4-byte port]
+    //   ...
+
+public:
+    static unsigned int
+    Serialize(const std::vector<std::map<PubKey, Peer>>& shards,
+              std::vector<unsigned char>& output, unsigned int cur_offset)
+    {
+        LOG_MARKER();
+
+        uint32_t numOfComms = shards.size();
+
+        // 4-byte num of committees
+        Serializable::SetNumber<uint32_t>(output, cur_offset, numOfComms,
+                                          sizeof(uint32_t));
+        cur_offset += sizeof(uint32_t);
+
+        LOG_GENERAL(INFO, "Number of committees = " << numOfComms);
+
+        for (unsigned int i = 0; i < numOfComms; i++)
+        {
+            const std::map<PubKey, Peer>& shard = shards.at(i);
+
+            // 4-byte committee size
+            Serializable::SetNumber<uint32_t>(output, cur_offset, shard.size(),
+                                              sizeof(uint32_t));
+            cur_offset += sizeof(uint32_t);
+
+            LOG_GENERAL(INFO,
+                        "Committee size = " << shard.size() << "\n"
+                                            << "Members:");
+
+            for (auto& kv : shard)
+            {
+                // 33-byte public key
+                kv.first.Serialize(output, cur_offset);
+                cur_offset += PUB_KEY_SIZE;
+
+                // 16-byte ip + 4-byte port
+                kv.second.Serialize(output, cur_offset);
+                cur_offset += IP_SIZE + PORT_SIZE;
+
+                LOG_GENERAL(
+                    INFO,
+                    " PubKey = "
+                        << DataConversion::SerializableToHexStr(kv.first)
+                        << " at " << kv.second.GetPrintableIPAddress()
+                        << " Port: " << kv.second.m_listenPortHost);
+            }
+        }
+
+        return cur_offset;
+    }
+};
+
+#endif // __DSBLOCKMESSAGE_H__
```

Comment:
Why file named "DSBlockMessage.h" contain the only one class named "ShardingStructure"? The naming seems no obvious relationship.

Reply:
When PoW merging is completed, the DS Block message will actually contain three things:
DS Block
Sharding structure
Txn sharing assignments

So, the naming is in preparation for the merging.
I'll put a comment in the Description section above to make this clear.